File: web/js/textnorm.js
/**
 * Text normalisation, fuzzy matching and unit parsing.
 *
 * A direct port of labelcheck/textnorm.py. The Python version remains the
 * reference implementation with its own test suite; tests/parity.mjs checks
 * that this port agrees with it, because two copies of compliance logic that
 * quietly disagree would be worse than having only one.
 */

export const ML_PER_FL_OZ = 29.5735295625;
export const ML_PER_PINT = 473.176473;
export const ML_PER_QUART = 946.352946;
export const ML_PER_GALLON = 3785.411784;

// Characters OCR commonly substitutes inside numbers.
const DIGIT_FIXES = {
  O: '0', o: '0', Q: '0', D: '0',
  l: '1', I: '1', '|': '1', i: '1',
  S: '5', s: '5', B: '8', Z: '2', z: '2', G: '6',
};

const COMPANY_STOPWORDS = new Set([
  'the', 'and', 'co', 'company', 'inc', 'incorporated', 'llc', 'llp', 'lp',
  'ltd', 'limited', 'corp', 'corporation', 'gmbh', 'sa', 'sas', 'spa',
  'srl', 'bv', 'nv', 'ag', 'pty', 'plc', 'of',
]);

export function stripAccents(value) {
  return String(value).normalize('NFKD').replace(/[̀-ͯ]/g, '');
}

export function normalize(value) {
  if (!value) return '';
  let text = stripAccents(String(value));
  text = text.replace(/&/g, ' and ');
  text = text.replace(/[^\w\s%./-]+/g, ' ');
  text = text.replace(/-/g, ' ').replace(/\//g, ' ').replace(/\./g, ' ');
  text = text.replace(/\s+/g, ' ');
  return text.trim().toLowerCase();
}

export function tokens(value) {
  const n = normalize(value);
  return n ? n.split(' ') : [];
}

export function significantTokens(value) {
  return tokens(value).filter((t) => !COMPANY_STOPWORDS.has(t));
}

/* ---- difflib.SequenceMatcher.ratio(), ported ---------------------------
 * Python's difflib uses the Ratcliff/Obershelp algorithm, not edit
 * distance. Substituting a Levenshtein ratio here would shift every
 * matching threshold, so the real algorithm is reproduced.
 */
function buildB2J(b) {
  const b2j = new Map();
  for (let i = 0; i < b.length; i += 1) {
    const ch = b[i];
    let list = b2j.get(ch);
    if (!list) { list = []; b2j.set(ch, list); }
    list.push(i);
  }
  return b2j;
}

function findLongestMatch(a, b2j, alo, ahi, blo, bhi) {
  let besti = alo;
  let bestj = blo;
  let bestsize = 0;
  let j2len = new Map();
  for (let i = alo; i < ahi; i += 1) {
    const newj2len = new Map();
    const positions = b2j.get(a[i]);
    if (positions) {
      for (const j of positions) {
        if (j < blo) continue;
        if (j >= bhi) break;
        const k = (j2len.get(j - 1) || 0) + 1;
        newj2len.set(j, k);
        if (k > bestsize) { besti = i - k + 1; bestj = j - k + 1; bestsize = k; }
      }
    }
    j2len = newj2len;
  }
  return [besti, bestj, bestsize];
}

function matchingTotal(a, b) {
  const b2j = buildB2J(b);
  let total = 0;
  const queue = [[0, a.length, 0, b.length]];
  while (queue.length) {
    const [alo, ahi, blo, bhi] = queue.pop();
    const [i, j, k] = findLongestMatch(a, b2j, alo, ahi, blo, bhi);
    if (k) {
      total += k;
      if (alo < i && blo < j) queue.push([alo, i, blo, j]);
      if (i + k < ahi && j + k < bhi) queue.push([i + k, ahi, j + k, bhi]);
    }
  }
  return total;
}

export function similarity(a, b) {
  if (!a || !b) return 0;
  if (a === b) return 1;
  const total = a.length + b.length;
  if (!total) return 1;
  return (2 * matchingTotal(a, b)) / total;
}

/** Best similarity between `needle` and any same-length run of `haystack`. */
export function bestWindowSimilarity(needle, haystack) {
  const needleN = normalize(needle);
  const hayN = normalize(haystack);
  if (!needleN || !hayN) return 0;
  if (hayN.includes(needleN)) return 1;

  const needleWords = needleN.split(' ');
  const hayWords = hayN.split(' ');
  if (!needleWords.length || !hayWords.length) return 0;

  const span = needleWords.length;
  let best = similarity(needleN, hayN);
  const widths = new Set([Math.max(1, span - 1), span, span + 1]);
  for (const width of widths) {
    if (width > hayWords.length) continue;
    for (let start = 0; start + width <= hayWords.length; start += 1) {
      const window = hayWords.slice(start, start + width).join(' ');
      const score = similarity(needleN, window);
      if (score > best) {
        best = score;
        if (best >= 0.999) return best;
      }
    }
  }
  return best;
}

/** Fraction of the needle's significant words present in the haystack. */
export function tokenCoverage(needle, haystack) {
  const needleTokens = significantTokens(needle);
  if (!needleTokens.length) return 0;
  const hayWords = tokens(haystack);
  if (!hayWords.length) return 0;
  const hayJoined = hayWords.join(' ');
  const haySet = new Set(hayWords);
  let found = 0;
  for (const token of needleTokens) {
    if (haySet.has(token)) {
      found += 1;
    } else if (token.length > 3 && hayWords.some((w) => similarity(token, w) >= 0.85)) {
      found += 1;
    } else if (token.length > 5 && hayJoined.replace(/ /g, '').includes(token)) {
      // Handles decorative letter-spacing: "B R O O K" -> "brook".
      found += 1;
    }
  }
  return found / needleTokens.length;
}

export function fieldMatchScore(expected, ocrText) {
  return Math.max(bestWindowSimilarity(expected, ocrText),
                  tokenCoverage(expected, ocrText));
}

export function fixOcrDigits(value) {
  return String(value).replace(/[OoQDlI|iSsBZzG]/g, (c) => DIGIT_FIXES[c] || c);
}

function toNumber(raw) {
  let cleaned = fixOcrDigits(String(raw).trim());
  cleaned = cleaned.replace(/(?<=\d),(?=\d)/g, '.');
  cleaned = cleaned.replace(/[^\d.]/g, '');
  if (!cleaned || (cleaned.match(/\./g) || []).length > 1) return null;
  const value = Number(cleaned);
  return Number.isFinite(value) ? value : null;
}

/* ---- Alcohol content --------------------------------------------------- */
const ALC_CONTEXT = /alc|abv|vol|alcohol|proof|grad/i;
const PERCENT_RE = /(?<![\d.])(\d{1,2}(?:[.,]\d{1,2})?)\s*(?:%|percent|per cent)/gi;
const PERCENT_LOOSE_RE =
  /(?:alc(?:ohol)?\.?\s*(?:by\s*vol(?:ume)?)?|abv)\s*[:. ]?\s*(?<![\d.])(\d{1,2}(?:[.,]\d{1,2})?)\b/gi;
const PROOF_RE =
  /(?<![\d.])(\d{1,3}(?:[.,]\d{1,2})?)\s*(?:°\s*)?proof|proof\s*[:. ]?\s*(?<![\d.])(\d{1,3}(?:[.,]\d{1,2})?)/gi;

export function parseAlcohol(text) {
  if (!text) return [];
  const results = [];
  const seen = new Set();

  const add = (kind, raw, source) => {
    const value = toNumber(raw);
    if (value === null) return;
    if (kind === 'abv' && !(value >= 0 && value <= 100)) return;
    if (kind === 'proof' && !(value >= 0 && value <= 200)) return;
    const key = `${kind}:${Math.round(value * 100) / 100}`;
    if (seen.has(key)) return;
    seen.add(key);
    results.push({
      kind,
      value,
      abv: kind === 'proof' ? value / 2 : value,
      source: String(source).trim(),
    });
  };

  for (const m of text.matchAll(PROOF_RE)) {
    const raw = m[1] || m[2];
    if (raw) add('proof', raw, m[0]);
  }
  for (const m of text.matchAll(PERCENT_RE)) {
    // Kept whether or not alcohol wording is nearby: some labels print only
    // "13.5%". ALC_CONTEXT is retained for readability of intent.
    void ALC_CONTEXT;
    add('abv', m[1], m[0]);
  }
  for (const m of text.matchAll(PERCENT_LOOSE_RE)) {
    add('abv', m[1], m[0]);
  }
  return results;
}

export function formatAlcohol(entry) {
  const trim = (n) => String(Number(n.toFixed(4)));
  if (entry.kind === 'proof') {
    return `${trim(entry.value)} proof (${trim(entry.abv)}% ABV)`;
  }
  return `${trim(entry.value)}% ABV`;
}

/* ---- Net contents ------------------------------------------------------ */
const UNIT_TO_ML = {
  ml: 1, milliliter: 1, millilitre: 1, milliliters: 1, millilitres: 1, cc: 1,
  l: 1000, liter: 1000, litre: 1000, liters: 1000, litres: 1000, lt: 1000,
  cl: 10, centiliter: 10, centilitre: 10,
  dl: 100,
  floz: ML_PER_FL_OZ, flozs: ML_PER_FL_OZ, fluidounce: ML_PER_FL_OZ,
  fluidounces: ML_PER_FL_OZ, oz: ML_PER_FL_OZ, ozs: ML_PER_FL_OZ,
  ounce: ML_PER_FL_OZ, ounces: ML_PER_FL_OZ,
  pt: ML_PER_PINT, pint: ML_PER_PINT, pints: ML_PER_PINT,
  qt: ML_PER_QUART, quart: ML_PER_QUART, quarts: ML_PER_QUART,
  gal: ML_PER_GALLON, gallon: ML_PER_GALLON, gallons: ML_PER_GALLON,
};

const QTY_RE = new RegExp(
  '(\\d{1,4}(?:[.,]\\d{1,3})?)\\s*'
  + '(fl\\.?\\s*oz|fluid\\s*ounces?|ounces?|oz|ml|milli ?lit(?:er|re)s?|'
  + 'cl|centilit(?:er|re)s?|dl|lit(?:er|re)s?|l|lt|pints?|pt|quarts?|qt|'
  + 'gallons?|gal|cc)\\b\\.?',
  'gi',
);

const BIG_UNITS = new Set(['pt', 'pint', 'pints', 'qt', 'quart', 'quarts',
                           'gal', 'gallon', 'gallons']);
const SMALL_UNITS = new Set(['floz', 'oz', 'ounce', 'ounces']);

function unitKey(raw) {
  return raw.toLowerCase().replace(/[^a-z]/g, '');
}

export function parseNetContents(text) {
  if (!text) return [];
  const raw = [];
  for (const m of text.matchAll(QTY_RE)) {
    const quantity = toNumber(m[1]);
    if (quantity === null || quantity <= 0) continue;
    const unit = unitKey(m[2]);
    const factor = UNIT_TO_ML[unit];
    if (factor === undefined) continue;
    raw.push({
      ml: quantity * factor,
      quantity,
      unit,
      source: m[0].trim(),
      start: m.index,
      end: m.index + m[0].length,
    });
  }

  // Merge compound imperial statements: "1 PT 9 FL OZ" is one quantity.
  const merged = [];
  let index = 0;
  while (index < raw.length) {
    const current = { ...raw[index] };
    let step = index + 1;
    while (step < raw.length
           && raw[step].start - current.end <= 2
           && BIG_UNITS.has(current.unit)
           && SMALL_UNITS.has(raw[step].unit)) {
      current.ml += raw[step].ml;
      current.source += ` ${raw[step].source}`;
      current.end = raw[step].end;
      step += 1;
    }
    merged.push(current);
    index = step;
  }

  return merged.map(({ start, end, ...rest }) => ({
    ...rest,
    ml: Math.round(rest.ml * 1000) / 1000,
  }));
}

export function formatMl(millilitres) {
  const trim = (n) => String(Number(n.toFixed(4)));
  if (millilitres >= 1000
      && Math.abs(millilitres / 1000 - Number((millilitres / 1000).toFixed(2))) < 1e-9) {
    return `${trim(millilitres / 1000)} L (${trim(millilitres)} mL)`;
  }
  return `${trim(millilitres)} mL`;
}

File: labelcheck/textnorm.py
"""Text normalisation, fuzzy matching and unit parsing.

OCR of stylised label artwork is noisy: serifs get read as punctuation,
zeros become letter O, and decorative kerning inserts spaces mid-word. Every
comparison in this tool therefore runs through normalisation first, and uses
windowed fuzzy matching rather than equality.

Standard library only - difflib gives us sequence similarity with no third
party dependency and no network access.
"""
import re
import unicodedata
from difflib import SequenceMatcher
from typing import Dict, List, Optional

# --- Unit conversions ------------------------------------------------------
ML_PER_FL_OZ = 29.5735295625
ML_PER_PINT = 473.176473
ML_PER_QUART = 946.352946
ML_PER_GALLON = 3785.411784

# Characters OCR commonly substitutes inside numbers.
_DIGIT_FIXES = str.maketrans({
    "O": "0", "o": "0", "Q": "0", "D": "0",
    "l": "1", "I": "1", "|": "1", "i": "1",
    "S": "5", "s": "5", "B": "8", "Z": "2", "z": "2", "G": "6",
})

_PUNCT_RE = re.compile(r"[^\w\s%./-]+", re.UNICODE)
_WS_RE = re.compile(r"\s+")

# Words that carry no distinguishing information when matching company names.
_COMPANY_STOPWORDS = {
    "the", "and", "co", "company", "inc", "incorporated", "llc", "llp", "lp",
    "ltd", "limited", "corp", "corporation", "gmbh", "sa", "sas", "spa",
    "srl", "bv", "nv", "ag", "pty", "plc", "of",
}


def strip_accents(value: str) -> str:
    decomposed = unicodedata.normalize("NFKD", value)
    return "".join(ch for ch in decomposed if not unicodedata.combining(ch))


def normalize(value: Optional[str]) -> str:
    """Case-fold, strip accents and punctuation, collapse whitespace."""
    if not value:
        return ""
    text = strip_accents(str(value))
    text = text.replace("&", " and ")
    text = _PUNCT_RE.sub(" ", text)
    text = text.replace("-", " ").replace("/", " ").replace(".", " ")
    text = _WS_RE.sub(" ", text)
    return text.strip().lower()


def tokens(value: Optional[str]) -> List[str]:
    normalized = normalize(value)
    return normalized.split() if normalized else []


def significant_tokens(value: Optional[str]) -> List[str]:
    """Tokens with corporate boilerplate removed, for company-name matching."""
    return [t for t in tokens(value) if t not in _COMPANY_STOPWORDS]


def similarity(a: str, b: str) -> float:
    """Similarity of two already-normalised strings, from 0.0 to 1.0."""
    if not a or not b:
        return 0.0
    if a == b:
        return 1.0
    return SequenceMatcher(None, a, b).ratio()


def best_window_similarity(needle: str, haystack: str) -> float:
    """Best similarity between `needle` and any same-length run of `haystack`.

    Whole-text similarity is useless here: a correct three-word brand name
    inside 400 words of OCR scores near zero. Instead we slide a window of
    roughly the needle's length across the haystack and keep the best score.
    """
    needle_n = normalize(needle)
    hay_n = normalize(haystack)
    if not needle_n or not hay_n:
        return 0.0
    if needle_n in hay_n:
        return 1.0

    needle_words = needle_n.split()
    hay_words = hay_n.split()
    if not needle_words or not hay_words:
        return 0.0

    span = len(needle_words)
    best = similarity(needle_n, hay_n)
    # Try windows a little shorter and longer than the needle, so a dropped
    # or hallucinated OCR word does not sink an otherwise good match.
    for width in {max(1, span - 1), span, span + 1}:
        if width > len(hay_words):
            continue
        for start in range(0, len(hay_words) - width + 1):
            window = " ".join(hay_words[start:start + width])
            score = similarity(needle_n, window)
            if score > best:
                best = score
                if best >= 0.999:
                    return best
    return best


def token_coverage(needle: str, haystack: str) -> float:
    """Fraction of the needle's significant words that appear in the haystack.

    Complements windowed similarity: it catches the case where a label
    reorders or line-breaks a company name, which shifts character
    positions but keeps every word.
    """
    needle_tokens = significant_tokens(needle)
    if not needle_tokens:
        return 0.0
    hay_words = tokens(haystack)
    if not hay_words:
        return 0.0
    hay_joined = " ".join(hay_words)
    found = 0
    for token in needle_tokens:
        if token in hay_words:
            found += 1
        elif len(token) > 3 and any(
            similarity(token, w) >= 0.85 for w in hay_words
        ):
            found += 1
        elif len(token) > 5 and token in hay_joined.replace(" ", ""):
            # Handles decorative letter-spacing: "B R O O K" -> "brook".
            found += 1
    return found / len(needle_tokens)


def field_match_score(expected: str, ocr_text: str) -> float:
    """Combined score used for free-text label fields."""
    return max(
        best_window_similarity(expected, ocr_text),
        token_coverage(expected, ocr_text),
    )


def fix_ocr_digits(value: str) -> str:
    """Repair letter-for-digit confusions inside an otherwise numeric token."""
    return value.translate(_DIGIT_FIXES)


def _to_float(raw: str) -> Optional[str]:
    cleaned = fix_ocr_digits(raw.strip())
    # European decimal comma, and OCR reading "." as ",".
    cleaned = re.sub(r"(?<=\d),(?=\d)", ".", cleaned)
    cleaned = re.sub(r"[^\d.]", "", cleaned)
    if not cleaned or cleaned.count(".") > 1:
        return None
    return cleaned


# --- Alcohol content -------------------------------------------------------
_ALC_CONTEXT = re.compile(
    r"alc|abv|vol|alcohol|proof|proof|grad", re.IGNORECASE
)
_PERCENT_RE = re.compile(
    r"(?<![\d.])(\d{1,2}(?:[.,]\d{1,2})?)\s*(?:%|percent|per cent)",
    re.IGNORECASE,
)
_PERCENT_LOOSE_RE = re.compile(
    r"(?:alc(?:ohol)?\.?\s*(?:by\s*vol(?:ume)?)?|abv)\s*[:. ]?\s*"
    r"(?<![\d.])(\d{1,2}(?:[.,]\d{1,2})?)\b",
    re.IGNORECASE,
)
_PROOF_RE = re.compile(
    r"(?<![\d.])(\d{1,3}(?:[.,]\d{1,2})?)\s*(?:°\s*)?proof|proof\s*[:. ]?\s*"
    r"(?<![\d.])(\d{1,3}(?:[.,]\d{1,2})?)",
    re.IGNORECASE,
)


def parse_alcohol(text: Optional[str]) -> List[Dict]:
    """Extract every alcohol-strength statement found in `text`.

    Returns dicts of {kind: 'abv'|'proof', value: float, abv: float,
    source: str}. `abv` is always populated, with proof halved, so callers
    can compare a proof-only label against an ABV-only application.
    """
    if not text:
        return []
    results: List[Dict] = []
    seen = set()

    def add(kind: str, raw: str, source: str):
        as_float = _to_float(raw)
        if as_float is None:
            return
        try:
            value = float(as_float)
        except ValueError:
            return
        abv = value / 2.0 if kind == "proof" else value
        # Reject impossible readings rather than reporting a false mismatch.
        if kind == "abv" and not (0.0 <= value <= 100.0):
            return
        if kind == "proof" and not (0.0 <= value <= 200.0):
            return
        key = (kind, round(value, 2))
        if key in seen:
            return
        seen.add(key)
        results.append({"kind": kind, "value": value, "abv": abv,
                        "source": source.strip()})

    for match in _PROOF_RE.finditer(text):
        raw = match.group(1) or match.group(2)
        if raw:
            add("proof", raw, match.group(0))

    for match in _PERCENT_RE.finditer(text):
        window = text[max(0, match.start() - 40):match.end() + 40]
        if _ALC_CONTEXT.search(window):
            add("abv", match.group(1), match.group(0))
        else:
            # A bare percentage with no alcohol context nearby is still worth
            # keeping as a weak candidate (some labels print just "13.5%").
            add("abv", match.group(1), match.group(0))

    for match in _PERCENT_LOOSE_RE.finditer(text):
        add("abv", match.group(1), match.group(0))

    return results


def format_alcohol(entry: Dict) -> str:
    if entry["kind"] == "proof":
        return f"{entry['value']:g} proof ({entry['abv']:g}% ABV)"
    return f"{entry['value']:g}% ABV"


# --- Net contents ----------------------------------------------------------
_UNIT_TO_ML = {
    "ml": 1.0, "milliliter": 1.0, "millilitre": 1.0, "milliliters": 1.0,
    "millilitres": 1.0, "cc": 1.0,
    "l": 1000.0, "liter": 1000.0, "litre": 1000.0, "liters": 1000.0,
    "litres": 1000.0, "lt": 1000.0,
    "cl": 10.0, "centiliter": 10.0, "centilitre": 10.0,
    "dl": 100.0,
    "floz": ML_PER_FL_OZ, "flozs": ML_PER_FL_OZ, "fluidounce": ML_PER_FL_OZ,
    "fluidounces": ML_PER_FL_OZ, "oz": ML_PER_FL_OZ, "ozs": ML_PER_FL_OZ,
    "ounce": ML_PER_FL_OZ, "ounces": ML_PER_FL_OZ,
    "pt": ML_PER_PINT, "pint": ML_PER_PINT, "pints": ML_PER_PINT,
    "qt": ML_PER_QUART, "quart": ML_PER_QUART, "quarts": ML_PER_QUART,
    "gal": ML_PER_GALLON, "gallon": ML_PER_GALLON, "gallons": ML_PER_GALLON,
}

# Matches "750 ML", "1.75 L", "12 FL. OZ.", "25.4 FLOZ", "1 PT 9 FL OZ".
_QTY_RE = re.compile(
    r"(\d{1,4}(?:[.,]\d{1,3})?)\s*"
    r"(fl\.?\s*oz|fluid\s*ounces?|ounces?|oz|ml|milli ?lit(?:er|re)s?|"
    r"cl|centilit(?:er|re)s?|dl|lit(?:er|re)s?|l|lt|pints?|pt|quarts?|qt|"
    r"gallons?|gal|cc)\b\.?",
    re.IGNORECASE,
)


def _unit_key(raw: str) -> str:
    return re.sub(r"[^a-z]", "", raw.lower())


def parse_net_contents(text: Optional[str]) -> List[Dict]:
    """Extract every net-contents statement, normalised to millilitres.

    Handles compound US statements such as "1 PT 9 FL OZ" by summing
    adjacent quantities that sit within a few characters of each other.
    """
    if not text:
        return []
    raw_matches = []
    for match in _QTY_RE.finditer(text):
        as_float = _to_float(match.group(1))
        if as_float is None:
            continue
        try:
            quantity = float(as_float)
        except ValueError:
            continue
        unit = _unit_key(match.group(2))
        factor = _UNIT_TO_ML.get(unit)
        if factor is None or quantity <= 0:
            continue
        raw_matches.append({
            "ml": quantity * factor,
            "quantity": quantity,
            "unit": unit,
            "source": match.group(0).strip(),
            "start": match.start(),
            "end": match.end(),
        })

    # Merge compound imperial statements: "1 PT 9 FL OZ" is one quantity.
    merged: List[Dict] = []
    index = 0
    while index < len(raw_matches):
        current = dict(raw_matches[index])
        step = index + 1
        while (
            step < len(raw_matches)
            and raw_matches[step]["start"] - current["end"] <= 2
            and current["unit"] in {"pt", "pint", "pints", "qt", "quart",
                                    "quarts", "gal", "gallon", "gallons"}
            and raw_matches[step]["unit"] in {"floz", "oz", "ounce", "ounces"}
        ):
            current["ml"] += raw_matches[step]["ml"]
            current["source"] += " " + raw_matches[step]["source"]
            current["end"] = raw_matches[step]["end"]
            step += 1
        merged.append(current)
        index = step

    for entry in merged:
        entry.pop("start", None)
        entry.pop("end", None)
        entry["ml"] = round(entry["ml"], 3)
    return merged


def format_ml(millilitres: float) -> str:
    """Render a millilitre value the way a label would."""
    if millilitres >= 1000 and abs(millilitres / 1000 - round(millilitres / 1000, 2)) < 1e-9:
        return f"{millilitres / 1000:g} L ({millilitres:g} mL)"
    return f"{millilitres:g} mL"
